lang/java/reef-io/src/main/java/org/apache/reef/io/network/group/impl/operators/ScatterReceiver.java

```diff
@@ -0,0 +1,164 @@
+                         @Parameter(OperatorName.class) final String operName,
+                         @Parameter(TaskConfigurationOptions.Identifier.class) final String selfId,
+                         @Parameter(DataCodec.class) final Codec<T> dataCodec,
+                         @Parameter(DriverIdentifier.class) final String driverId,
+                         @Parameter(TaskVersion.class) final int version,
+                         final CommGroupNetworkHandler commGroupNetworkHandler,
+                         final NetworkService<GroupCommunicationMessage> netService,
+                         final CommunicationGroupServiceClient commGroupClient) {
+    LOG.finest(operName + "has CommGroupHandler-" + commGroupNetworkHandler.toString());
+    this.version = version;
+    this.groupName = Utils.getClass(groupName);
+    this.operName = Utils.getClass(operName);
+    this.dataCodec = dataCodec;
+    this.topology = new OperatorTopologyImpl(this.groupName, this.operName,
+                                             selfId, driverId, new Sender(netService), version);
+    this.commGroupClient = commGroupClient;
+    commGroupNetworkHandler.register(this.operName, this);
+  }
+
+  @Override
+  public int getVersion() {
+    return version;
+  }
+
+  @Override
+  public void initialize() throws ParentDeadException {
+    topology.initialize();
+  }
+
+  @Override
+  public Class<? extends Name<String>> getOperName() {
+    return operName;
+  }
+
+  @Override
+  public Class<? extends Name<String>> getGroupName() {
+    return groupName;
+  }
+
+  @Override
+  public String toString() {
+    final StringBuilder sb = new StringBuilder("ScatterReceiver:")
+        .append(Utils.simpleName(groupName))
+        .append(":")
+        .append(Utils.simpleName(operName))
+        .append(":")
+        .append(version);
+    return sb.toString();
+  }
+
+  @Override
+  public void onNext(final GroupCommunicationMessage msg) {
+    topology.handle(msg);
+  }
+
+  @Override
+  public List<T> receive() throws NetworkException, InterruptedException {
+    LOG.entering("ScatterReceiver", "receive");
+    // I am intermediate node or leaf.
+    LOG.fine("I am " + this);
+
+    if (init.compareAndSet(false, true)) {
+      LOG.fine(this + " Communication group initializing.");
+      commGroupClient.initialize();
+      LOG.fine(this + " Communication group initialized.");
+    }
+
+    try {
+      LOG.fine(this + " Waiting to receive scatter from parent.");
+      final byte[] data = topology.recvFromParent(ReefNetworkGroupCommProtos.GroupCommMessage.Type.Scatter);
+
+      if (data == null) {
+        LOG.fine(this + " Received null. Perhaps one of my ancestors is dead.");
+        LOG.exiting("ScatterSender", "receive", null);
+        return null;
+      }
+
+      LOG.fine(this + " Successfully received scattered data.");
+      final Pair<byte[][], Map<String, byte[]>> myDataAndChildBytesMap = ScatterHelper.decodeScatterMsgForNode(data);
```

Comment:
Feels like this should be a separate class. I think it will also simplify ScatterHelper part. I would have this object and a codec for it instead of the helper.